[MRG] Simplify raw_to_bids

jasmainak opened the pull request:
closes #98 
closes #105
closes #119 
closes #123

jasmainak commented on examples/convert_mne_sample.py:
```diff
@@ -29,17 +30,18 @@
-            overwrite=True)
+raw = read_raw(fname)
+bids_fname = make_bids_filename(subject='01', run='01', session='01',
+                                task='audiovisual', suffix='meg.fif')
```
maybe we don't need `kind` argument in `write_raw_bids` if we do this ...

sappelhoff replied:
mmh why wouldn't we? The writing of some fields such as `EEGReference` is still contingent on the "kind" of data.

jasmainak replied:
Because it's in the "suffix" when you provide the `bids_fname` ...

teonbrooks replied:
i would prefer to `kind` over suffix. Suffix implies to the user that if I add another file extension, MNE is able to say take a Neuroscan `.cnt` and output an `.edf` and that is something we are not capable of. We have worked on the underlying principle that you get the same file format out that you started with

jasmainak replied:
okay I could add a check so that the extension does not change? If you don't give the filename extension, you can't construct a full filename ...

jasmainak replied:
@teonbrooks how do you propose to resolve this? Should we have two arguments `kind` and `extension` ? Or do you say the extension can go in `suffix`?

teonbrooks replied:
i can't get past explicitly specifying the modality in the suffix but it being determined automatically in the write function. i'm -1 on the way a filename is constructed.

jasmainak replied:
we don't have to determine the modality automatically if we provide it in the filename ...

jasmainak replied:
@teonbrooks do you have a suggestion for better API?

jasmainak replied:
I thought this through once more. So far, we have been using "suffix" in our internal API to construct filenames. See for instance [here](https://github.com/mne-tools/mne-bids/blob/master/mne_bids/mne_bids.py#L634), [here](https://github.com/mne-tools/mne-bids/blob/master/mne_bids/mne_bids.py#L658), and [here](https://github.com/mne-tools/mne-bids/blob/master/mne_bids/mne_bids.py#L662). My feeling is that we should not have different conventions in our internal API vs. our external API -- so I am actually okay leaving it as `suffix`. We can't have both `suffix` and `kind` to construct the last part of the filename because that would mean we have two ways of doing the same thing (goes against Zen of Python). #105 may be redundant after all or we could rewrite it to do some checks ...

sappelhoff replied:
okay from my side.

jasmainak replied:
okay last commit should also take care of #105

jasmainak commented on mne_bids/mne_bids.py:
```diff
@@ -485,30 +484,20 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
-                verbose=True):
+def write_raw_bids(raw, bids_fname, output_path, kind='meg', events_data=None,
+                   event_id=None, eeg_ref=None, eeg_gnd=None, overwrite=True,
+                   verbose=True):
```
I'm imagining an analogous function called `read_raw_bids` for our `0.2` release

monkeyman192 commented on mne_bids/mne_bids.py:
```diff
@@ -485,30 +484,20 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
-        Acquisition parameter for the dataset.
-    run : int | None
-        The run number for this dataset.
+    bids_fname : str
```
I feel like this description doesn't really capture what you are expecting the user to enter

sappelhoff replied:
+1 this is also problematic with regards to `copyfile_brainvision` and `copyfile_eeglab`, see below

jasmainak replied:
okay I changed it. Let me know if it looks better to you

monkeyman192 replied:
Looks better, but I would still throw in "expected" or "to be produced" or something to indicate that it is the file name of a file that doesn't exist before export.

jasmainak commented:
@sappelhoff @teonbrooks @monkeyman192 thoughts?

I stop for today. The diff is a bit largish but I _think_ I got two examples to run. We should first agree that this is a more future-compatible / simpler API. Maybe it isn't ... I don't know

jasmainak commented:
and your turn to review :-) I take a break from reviewing ...

jasmainak commented on doc/api.rst:
```diff
@@ -13,7 +13,17 @@ MNE BIDS (:py:mod:`mne_bids.mne_bids`):
 
-   raw_to_bids
+   write_raw_bids
+   make_bids_filename
```
the older API was approaching monolithic design. See [unix philosophy](https://en.wikipedia.org/wiki/Unix_philosophy#Eric_Raymond%E2%80%99s_17_Unix_Rules) for inspiration on monolithic vs. modular design.

sappelhoff commented on mne_bids/mne_bids.py:
```diff
@@ -594,18 +565,6 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
     data_meta_fname = make_bids_filename(
```
`data_meta_fname` should be `sidecar_fname` in my opinion

monkeyman192 replied:
+1
I feel this should have been renamed a while ago...

jasmainak replied:
Let's do this elsewhere. It's not the purpose of this pull request ...

jasmainak replied:
Actually I addressed this so we can focus on more important design issues :)

sappelhoff commented on mne_bids/mne_bids.py:
```diff
@@ -594,18 +565,6 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
-            subject=subject_id, session=session_id, task=task, run=run,
-            acquisition=acquisition, suffix='%s%s' % (kind, ext),
-            prefix=raw_folder)
     events_tsv_fname = make_bids_filename(
```
`events_tsv_fname` should be `events_fname` ... or else all other `XXX_fname` should be `XXX_tsv_fname`/`XXX_json_fname`

jasmainak replied:
same comment as above. Let's make one PR do one thing ...

jasmainak replied:
okay done anyway :)

sappelhoff commented on mne_bids/utils.py:
```diff
@@ -60,6 +60,20 @@ def _mkdir_p(path, overwrite=False, verbose=False):
             raise
 
 
+def _parse_bids_filename(fname):
```
you need a test ;-)

jasmainak replied:
typically private functions are not tested explicitly. But these lines should be covered since they are being called by the `write_raw_bids` function

sappelhoff commented on mne_bids/mne_bids.py:
```diff
@@ -657,17 +613,17 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
     # BrainVision is multifile, copy over all of them and fix pointers
     elif ext == '.vhdr':
-        copyfile_brainvision(raw_fname, raw_file_bids)
+        copyfile_brainvision(raw_fname, fname)
```
this will not go well together with what you write in the docstring on line 497:
```
The file path of the BIDS compatible raw file. In the case of multifile
        systems (e.g., vhdr, .ds etc.), this is the path to a folder containing
        all the files.
```

See also my comment above. Both `copyfile_brainvision` (a multifile system, always 3 files) and `copyfile_eeglab` (a facultative multifile system ... can be 1 file or 2 files) expect the location of the header file, i.e., `myfile.vhdr` or `myfile.set` ... passing a directory is currently not supported and would maybe also not be a good idea.

jasmainak commented on mne_bids/mne_bids.py:
```diff
@@ -551,26 +528,23 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
+    raw_fname = raw.filenames[0]
+    if '.ds' in op.dirname(raw.filenames[0]):
+        raw_fname = op.dirname(raw.filenames[0])
+    _, ext = _parse_ext(raw_fname, verbose=verbose)
```
@sappelhoff I need your help here. For `.vhdr` files, `raw.filenames[0]` has the extension `.eeg` and not `.vhdr`. Is this expected?

sappelhoff replied:
yes, this is how MNE-Python handles the `filenames` attribute. It places the datafile there, which in the case of BrainVision is the `.eeg`.

jasmainak commented:
CIs are happy

teonbrooks commented on examples/convert_group_studies.py:
```diff
@@ -63,13 +65,15 @@
-                    event_id=event_id, output_path=output_path, overwrite=True)
+        raw_fname = op.join(data_path, repo, subject, 'MEG',
+                            'run_%02d_raw.fif' % run)
+        raw = mne.io.read_raw_fif(raw_fname)
```
is there a reason to create a Raw object as opposed to just providing the path above

jasmainak replied:
Two reasons for me:

1. `raw` object can be manipulated. If you just give a filename, users will start asking us for functions or new arguments for every small change they need to do. Instead, they could just change the `raw` object with one or two lines of `numpy` code when necessary
2. The `read_raw` object is likely to grow larger with more arguments as we incorporate more file types. It just feels to me that a general `raw` reader should be kept separate from the issue of writing out a BIDS compatible folder.

teonbrooks commented:
so from what I gathered, the proposed api would be:

1. create Raw object with an exposed general `read_raw` function
 - this gets all the auxiliary files together and have them unified in the Raw obj
2. generate a filename?
3. write out bids directory, raw file, and sidecar files

I don't think the 2nd step is necessary. I feel that this should be handled in the `write_raw_bids`

monkeyman192 commented:
>this gets all the auxiliary files together and have them unified in the Raw obj

by making everything a single raw object we lose extra files that were used to construct it (eg, for KIT data the .elp, .hsp and .mrk files which all also need to be copied to the BIDS directory)

monkeyman192 commented:
I also still dislike the fact that you have to create the bids file name to pass it in as an argument. But then the name gets deconstructed down to the constituent parts which are then used to remake the name.
I think I have said it before but why not just have essentially a dictionary to pass the data in.
Or create a `BIDSParameters` (eg.) object which contains all the parameters. Just something so you don't need to construct then deconstruct the name.

jasmainak commented:
> 2. generate a filename?
> 3. write out bids directory, raw file, and sidecar files
> I don't think the 2nd step is necessary. I feel that this should be handled in the write_raw_bids

hmm ... it feels a bit more natural to me to give a filename to a `write_raw_bids` function. But, I'm willing to consider alternative APIs. 

I wonder what opinion folks with more experience in API design (like @dengemann or @agramfort ) might have.

monkeyman192 commented:
> so far I'm able to make all the tests work

That's because this functionality is added in #114 and #77 :)

> not sure ... this info is not anywhere in the raw.info object?

I am pretty sure the only file path stored is the main file. Any extra files are discarded once their information has been absorbed into the raw object. This is mainly an issue for KIT data, but BTi data can also provide a headshape file...)

jasmainak commented on mne_bids/mne_bids.py:
```diff
@@ -382,11 +381,6 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
-        Description of the type of reference used and (when applicable) of
-        location of the reference electrode.  Defaults to None.
-    eeg_gnd : str
-        Description  of the location of the ground electrode. Defaults to None.
```
@sappelhoff I removed these two because I feel it's something that the user should do directly via a text editor, not python. 

I would say let's not aim for perfection, but if we can get 80% of the job done, `mne-bids` will be useful for most people.

sappelhoff replied:
okay, fair enough for `eeg_gnd`, but `eeg_ref`(which is EEGReference in BIDS) is a REQUIRED field in the EEG BEP006, so I would keep it.

jasmainak replied:
alright done

jasmainak commented on mne_bids/mne_bids.py:
```diff
@@ -659,7 +646,7 @@ def write_raw_bids(raw, bids_fname, output_path, events_data=None,
         sh.copyfile(raw_fname, bids_fname)
     # KIT data requires the marker file to be copied over too
-    if hpi is not None:
+    if raw.info['dig'] is not None:
```
@monkeyman192 I started rewriting this but I think I need your opinion before I jump right in. Why is the `_marker` file for KIT data not stored in the same folder as the raw data -- like in the case of CTF? I think MNE-BIDS is not for the case where files are scattered all around. It's more when files are already semi-organized but it's not necessarily BIDS format.

What we could do is to reverse engineer the [read_mrk](https://github.com/mne-tools/mne-python/blob/e4c995b4517089f7f105a9afb997f1b05a4e3653/mne/io/kit/coreg.py#L16) function that @teonbrooks wrote in MNE and write out the `sqd` marker file from that. We do not need to worry about what format the file was originally in because as we were discussing in the BIDS specification github page, only one filetype per manufacturer is going to be supported for these kinds of files.

Let me know what you think.

monkeyman192 replied:
the marker file is stored in the same location...
The sub folder with the same name as the sidecar.json contains the .mrk and .con file:
![image](https://user-images.githubusercontent.com/8000597/47050613-c9078380-d1ec-11e8-9cf3-038d4cfa7eef.png)
or do you mean before exporting?
Generally the .mrk file *should* be in the same folder as the .con file, but it doesn't have to, nor will it necessarily have the same name.

Also, doing the reverse of that function you will lose a vast amount of data. As you can see in that picture above the .mrk files we have are ~3Mb, even though most of it isn't used.

jasmainak replied:
Hmm okay I see. If it’s KIT specific issue, we should probably keep it out of MNE-BIDS for now then. We should go by 80-20 philosophy.

jasmainak replied:
> since not being able to handle the markers is the same as not being able to handle the headshape/electrode placement files also

we won't have this issue of losing information with the headshape files, no? They are simple text files and we can write them out. This is only an issue with `hpi` as far as I understand.

monkeyman192 replied:
Ah yeah, that's true. Yeah, hpi only seems to be used for the KIT data also. Really I feel like it should be something that is stored in the Raw object and extracted if need-be (like even just store the path in the `raw.filenames` property).
If the value was stored here, then this would become a non-issue.

jasmainak replied:
I mean you could in principle store it as a private attribute of `raw`. I think nothing stops us from doing that in our `read_raw` function. But it feels a bit of a hack to me, so I'd wait a bit before going that way ...

monkeyman192 replied:
yeah, that's why I propose adding it to the `raw.filenames` property. It's already a list, and for KIT data at least the .mrk file is necessary. It doesn't seem *too* unreasonable for the marker file to be listed there too...
You could also maybe just add it to the `raw._raw_extras` dictionary.

jasmainak replied:
yes, I agree that makes sense ...

teonbrooks replied:
`hpi` is an argument of `read_raw` function. the bidsification could happen there. I think it is a blocker if we don't accept hpi files because that breaks pipelines for source modeling for non FIF and CTF systems

jasmainak replied:
alright fixed. It works now as long as you use our general reader :)

jasmainak commented:
PR is rebased. Reviews welcome. 

There is the outstanding issue of `suffix` but is there any suggestion for alternate API? In the meanwhile, I'll also sleep over it ...

monkeyman192 commented:
Can you not just leave it out? Why do you need the kind there? Once #105 goes through we don't have a need for it. Although upon comparing this and #105, I wonder how you plan on making them compatible?

So for example, how are you meant to specify the kind as `'auto'` in this? You either need to have it all just happen automatically (ie. no `kind` argument anywhere and just let it do it's thing), or give `write_raw_bids` a `kind` argument.

jasmainak commented:
No, we can't leave it out because the filename would be incomplete in that case. We need a filename that looks complete -- such as "sub-01_ses-01_run-01_task-auditory_meg.fif". If we leave out `'kind'`, then we won't have a full filename. With this API, I think many people wouldn't even need `make_bids_filename` necessarily as they could just use the string directly.

Yes, it's not so easy to reconcile this PR with #105. This is why I am inclined to tackle this PR before #105. I think we could still use the code of #105 to do validation. That is, whether the kind/suffix we supplied makes sense for the `raw` object we have. It does mean we would have to supply kind/suffix explicitly when creating the filename but I think this is what @sappelhoff and @teonbrooks preferred anyway.

jasmainak commented:
All checks passed. Hurray!

monkeyman192 commented on examples/convert_group_studies.py:
```diff
@@ -63,13 +65,15 @@
-                    output_path=output_path, overwrite=True)
+        raw_fname = op.join(data_path, repo, subject, 'MEG',
+                            'run_%02d_raw.fif' % run)
+        raw = mne.io.read_raw_fif(raw_fname)
```
I think this should use `read_raw` for the reading of the raw file.
Because KIT data *has* to use the `read_raw` function (a point that should probably be mentioned somewhere) I think it would be good to have the standard that we use `read_raw` everywhere to avoid any issues. (see other comments re this point...)

jasmainak commented on mne_bids/tests/test_mne_bids.py:
```diff
@@ -242,60 +250,59 @@ def test_edf():
-                output_path=output_path, overwrite=False, kind='eeg')
+    bids_fname = partial_bids_fname + '_eeg.set'
+    raw = read_raw(raw_fname)
+    raw._filenames = [raw_fname]
```
The reason for removing the hack is that with the new API, `mne_bids` absolutely requires `raw.filenames` to exist as otherwise, it doesn't know where the original files are stored. I think this is a reasonable assumption to make since you know the path when reading in the file -- so if it's missing, that's more a bug upstream in MNE that should be fixed. It does mean `.set` files cannot be converted with MNE development version but I can live with that for a couple of months ...

jasmainak commented:
comments addressed @monkeyman192 @sappelhoff !

sappelhoff commented on mne_bids/tests/test_mne_bids.py:
```diff
@@ -54,10 +59,10 @@ def test_fif():
-                acquisition=acq, task=task, raw_file=raw_fname,
-                events_data=events_fname, output_path=output_path,
-                event_id=event_id, overwrite=False)
+    raw = mne.io.read_raw_fif(raw_fname)
```
+1

jasmainak replied:
see discussion [above](https://github.com/mne-tools/mne-bids/pull/106/files#r226822540). Resolving this conversation in favor of that one.

sappelhoff commented on mne_bids/mne_bids.py:
```diff
@@ -528,56 +514,31 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
-    task : str
-        Name of the task the data is based on.
-    raw_file : str | instance of mne.Raw
+    raw : instance of mne.Raw
```
+1 ... we do have some special code in our `read_raw` that can help users to not run into errors

jasmainak replied:
I don't want to necessarily advocate strongly for that approach of using `read_raw`. There have been many discussions in the past on the MNE github list about the possibility of a generalized reader (like `scipy.misc.imread`) but most people were in general -1 on this idea. I wasn't convinced at the moment but I've come around to the idea that `mne.io.read_raw_xxx` exists for a good reason. We can leave it in there to see how many people are using it as it's a thin wrapper over the `mne.io.read_raw_xxx` functions but I don't want to force people to use it. 

I've now added back the `hpi` parameter in the `write_raw_bids` function for now as otherwise there was this coupling between the two functions (`write_raw_bids` and `mne_bids.io.read_raw`) which was a bit odd. Might be worth waiting for @teonbrooks comments before merging this PR though as the generalized reader was his idea.

monkeyman192 replied:
The only reason we would need it here is because of the HPI issue with KIT data. If this functionality was moved into the MNE code then you could go back to to using the specific functions with no worries. If people think that the change made here is good then really I think it should be implemented in MNE that way there is no difference between KIT raw files in MNE-BIDS and MNE-python

jasmainak replied:
yes, there is probably something to fix upstream in MNE but so far it seems like an isolated case for KIT data. I can live with one extra parameter for that. We can deprecate it later if we come up with a better solution. But I think the raw files from `mne_bids.io.read_raw` and `mne.io.read_raw_xxx` should definitely be identical.

sappelhoff commented:
changes approved from my side :-) two more comments by @monkeyman192 to address, then we can merge.

teonbrooks commented:
I think the suffix argument should be removed and the filename generate function `make_bids_filename` should be renamed `make_bids_basename`. Here's why:

- As a user, having a suffix argument might lead you to think that you can convert your filetype say from `.set` to `.edf` when you change the suffix. This is also another way to introduce unforeseen errors, all that can be mitigated from the input file. I think it is too flexible.
- The filename extension should come from reading the data file in `read_raw` or it should come from the `Raw` object in `raw.filename` if that is being used.
- Requiring the modality in the suffix to be specified essentially just replaces the `kind` argument with the suffix argument. This would kill the auto detect functionality you proposed in #105.

teonbrooks commented:
also, we need to address the renaming of hpi (MEG-space marker) files, electrode (digitizer-space marker) files, hsp (digitizer-space) files. currently for files not stored as `.fif`, these auxiliary files are not being renamed and added to the folder.

jasmainak commented:
> also, we need to address the renaming of hpi (MEG-space marker) files, electrode (digitizer-space marker) files, hsp (digitizer-space) files. currently for files not stored as .fif, these auxiliary files are not being renamed and added to the folder.

we addressed this already. We have an argument for `hpi` in `write_raw_bids` ...

jasmainak commented:
> I think the suffix argument should be removed and the filename generate function make_bids_filename should be renamed make_bids_basename

what would `make_bids_basename` generate? Everything up to before `_meg.fif` ? And then you are proposing having a `kind`argument for the rest? I'm a little unclear on this proposal ...

> As a user, having a suffix argument might lead you to think that you can convert your filetype say from .set to .edf when you change the suffix. This is also another way to introduce unforeseen errors, all that can be mitigated from the input file. I think it is too flexible.

okay but this is easily fixed using an explicit error message. Let me do that.

> Requiring the modality in the suffix to be specified essentially just replaces the kind argument with the suffix argument. This would kill the auto detect functionality you proposed in #105.

yes, see above. This PR closes #105. See [this](https://github.com/mne-tools/mne-bids/pull/106/commits/e1386b8c2327b044c4897c44d0fb63abecf95008) commit which essentially replaces the functionality there. And the code is much simpler.

teonbrooks commented on mne_bids/commands/mne_bids_raw_to_bids.py:
```diff
@@ -35,6 +36,9 @@ def run():
+    parser.add_option('--acq', dest='acq',
+                      help='The acquisition parameter.',
+                      metavar='acq')
     parser.add_option('--kind', dest='kind',
```
`kind` is still in the cli tool

jasmainak replied:
I didn't change how the CLI works. It should work just as before ...

teonbrooks replied:
it's mostly a concern over unified api experience. after talking offline, I will wait til your next commit(s) to comment further

jasmainak replied:
yeah I see. I'm not sure how to expose a modular API for CLI though :) It's kind of similar for MNE report though. The Python API is really nice but the CLI is just a beast with many parameters. There are ways of chaining commands too but I think that would be longer discussion + bigger diff + more work etc. I would leave it out of the scope of this PR and release

teonbrooks commented on mne_bids/mne_bids.py:
```diff
@@ -105,14 +106,9 @@ def _channels_tsv(raw, fname, overwrite=False, verbose=True):
 
     # get the manufacturer from the file in the Raw object
     manufacturer = None
-    if hasattr(raw, 'filenames'):
```
is this fixed? I see that it is removed but doesn't really relate with the pr topic

jasmainak replied:
It does because now it should raise an error if `raw.filenames` does not exist ...

teonbrooks commented:
just discussed with @agramfort about the pr and he made this suggestion:

1. we can keep the generic read_raw functionality that loads all the info and files for the Raw object.
2. we should create a PR against `mne-python` such that headshape filenames, hpi filenames, and electrode filenames are saved as attributes when loading a raw file e.g. `raw.hpi`, `raw.hsp`, `raw.electrode`.
3. we can then copy the files with the proper BIDS formatting into the respective folders.

mock api:
```python
from mne_bids import read_raw

raw = read_raw(rawfile, hsp='string_path', electrode='string_path', hpi='string_path')
write_raw_bids(raw, subject, task, session, run)
```

the filename would not need to be constructed before writing the files into the folder, we reduce the friction for loading and writing for the user.

jasmainak commented:
Just posting some more comments/thoughts from @agramfort below for archiving:

 - `write_raw_bids` should work only for `fif` data because we cannot guarantee that the `raw` object has not changed significantly. So, now if you copy the original files, they will not work.
- one possibility is to pass kwargs which triages based on the file extension and retain `raw_to_bids`. Then `_read_raw` would remain private. However, we should then rename `raw_to_bids` as it would not accept `raw` anymore? Any suggestions?

comment from @jasmainak + @agramfort : `make_bids_filename` should probably not exist as it reinvents the wheel what can be done using string formatting.

sappelhoff commented:
> write_raw_bids should work only for fif data because we cannot guarantee that the raw object has not changed significantly. So, now if you copy the original files, they will not work.

Could you elaborate on this point? What does it mean? That we drop all other data formats? That we do write all BIDS files, but do not copy over the raw data files with changed name (e.g., like `copyfile_brainvision`). Also: I am not convinced of the argument that the raw object might have changed along the way --> where and how could it be changed if we read it from disk, do nothing, and then write it to disk?

sappelhoff commented:
> one possibility is to pass kwargs which triages based on the file extension and retain raw_to_bids. Then _read_raw would remain private. However, we should then rename raw_to_bids as it would not accept raw anymore? Any suggestions?

What kind of kwarg are you targeting? Examples please :-)

teonbrooks commented:
@sappelhoff I think that there was some confusion. I think the consideration is that if raw is a string, it remains the same format as the original, it is just copied in placed in the bids structure. if the raw is a Raw object, then it is written as a fif file because the user may have modified the information related to the file and FIF is the only format we write out. this is subject to change if the edf writer or brainvision writer ever land and we support it.

teonbrooks commented:
closing this in favor of continuing the discussion at #123

jasmainak commented:
> Could you elaborate on this point? What does it mean? That we drop all other data formats? That we do write all BIDS files, but do not copy over the raw data files with changed name (e.g., like copyfile_brainvision)?

no I don't think so. I think it's more an argument of weather we should allow users to access the `raw` object because they are then free to change it if they like. Which could lead to unexpected mismatch between the data we copy over from the original directories and the (modified) metadata in `raw.info`. Hence the argument to save to `fif` if it's a `raw` object because MNE can write to `fif` format but not others.

jasmainak commented:
> the filename would not need to be constructed before writing the files into the folder, we reduce the friction for loading and writing for the user.

@teonbrooks just putting out some more thoughts on this. If you use `make_bids_filename`, it's definitely more friction I agree. However, if you see the "quickstart" example, it's just a plain string like `'sub-01_run-02_task-audiovisual'` which is certainly less friction than specifying `raw_to_bids(..., subject_id='01', run='02', task='audiovisual')`. For me, `make_bids_filename` is really only for educational purposes -- I see most people just using plain old string formatting where needed. 

By the way, as we speak, there is talk of adding another [key-value pair](https://github.com/bids-standard/bids-specification/pull/62) to the spec. This would lead to yet another argument in the function ...

jasmainak commented on examples/convert_eeg_to_bids.py:
```diff
@@ -87,7 +89,11 @@
-raw = read_raw_edf(edf_path, preload=True)
+
+raw = read_raw_edf(edf_path, preload=False, stim_channel=None)
+annot = read_annotations_edf(edf_path)
```
@agramfort this should not be necessary in MNE ... `read_raw_edf` should work with `preload=False` since you have all the necessary components.

jasmainak replied:
also I'm a bit confused about what namespace `read_annotations_xxx` belong to. You have `mne.read_annotations` for fif files and then you have `mne.io.read_annotations_brainvision` and `mne.io.read_annotations_eeglab`. And for edf it is `mne.io.edf.edf.read_annotations_eeglab` ...

jasmainak replied:
oops, looks like docstring of this function is [also incorrect](https://github.com/mne-tools/mne-python/blob/master/mne/io/edf/edf.py#L1127)

agramfort replied:
yes agreed. We are not consistent. Can you open an issue on mne-python tracker to discuss this?

jasmainak replied:
done!

jasmainak commented on examples/convert_eeg_to_bids.py:
```diff
@@ -155,21 +164,18 @@
-                    event_id=trial_type,
-                    overwrite=True
-                    )
+        edf_path = eegbci.load_data(subject=subj_idx, runs=task_idx)[0]
```
@sappelhoff I changed this a bit. I find this alternative more readable. Hope this is okay for you.

sappelhoff replied:
+1 generally, however specifically I find it more readable to have each argument to a function on its own line. But that's just my opinion about this thing ... the remaining changes are fine.

jasmainak commented:
Okay I updated this PR after recent discussions with @agramfort and @teonbrooks 

- API is still `write_raw_bids(raw, bids_fname, ...)`
- However, we now have protection mechanism to ensure data cannot be modified (`raw.preload=False`)

Few caveats though:
- `mne.io.read_raw_edf` doesn't work smoothly with `preload=False`. But this should  be fixed upstream.
- `mne.io.read_raw_eeglab` doesn't work for `preload=False` if all the data is in `.set` file (instead of `.fdt`). Is this a blocker for you @sappelhoff ?
- The user can technically still crop the `raw` file even if `preload=False`. Unfortunately, MNE does not appear to keep track of cropping. The simple fix would be to add a boolean `raw._is_cropped` to MNE and use that to safeguard our conversion.
- We would need to use `0.17dev0` for the `edf` annotation reader and for `raw._is_cropped`. This means we would recommend dev version of MNE for a while until `0.17` is released.

sappelhoff commented:
> mne.io.read_raw_eeglab doesn't work for preload=False if all the data is in .set file (instead of .fdt). Is this a blocker for you @sappelhoff ?

No it's not a blocker, because it's just a secondary data format ("accepted but not recommended") for both EEG and iEEG (and personally, I don't like it 😉 ... bad documentation and based on Matlab).

However the flip side is that we'd need to implement preload=False capability for .set files in MNE-Python. This might be a work intensive PR ... and perhaps entirely impossible for old .set files (=old mat files).

For new mat files, it should be possible to load only selective fields into memory and leave the "data" untouched (preload=False), see [here](https://de.mathworks.com/help/matlab/import_export/load-parts-of-variables-from-mat-files.html)

agramfort commented on doc/api.rst:
```diff
@@ -23,10 +25,9 @@ Utils (:py:mod:`mne_bids.utils`):
-   make_bids_filename
    make_dataset_description
    copyfile_brainvision
+   copyfile_eeglab
```
do you need to make these functions public?

sappelhoff replied:
at least the `copyfile_brainvision` function is a feature that we also covered in an example: https://mne-tools.github.io/mne-bids/auto_examples/rename_brainvision_files.html#sphx-glr-auto-examples-rename-brainvision-files-py

jasmainak replied:
I guess these are useful functions? I don't know ...

teonbrooks replied:
I don't think they fit on the api page, but I think they should remain public through the utils

jasmainak replied:
But we have an example using these functions. Do you propose to remove the example too?

agramfort commented on mne_bids/commands/mne_bids_raw_to_bids.py:
```diff
@@ -4,11 +4,11 @@
 
 example usage:  $ mne_bids raw_to_bids --subject_id sub01 --task rest
---raw_file data.edf --output_path new_path
+--raw_fname data.edf --output_path new_path
```
I think --raw would be enough.

--raw is used in "mne browse_raw"

agramfort commented:
if you think something should be added/improved in mne-python please open
an issue to discuss it.

jasmainak commented:
@agramfort added check for cropping. The check won't work for BTi data unfortunately because I don't have all the info in the `raw` object to read back the file. Otherwise, addressed your comments.

agramfort commented:
can you submit the PR to mne-python to store as attribute in the init
what you miss?

jasmainak commented:
should I just add this param or should we store all the params that we use to read the raw object?

agramfort commented:
just put as object attribute all the init params you need to recreate the
raw instance from scratch.
it does not matter if it's not stored to disk as we don't write such files
anyway.

jasmainak commented:
@agramfort @teonbrooks CIs are happy

teonbrooks commented on README.md:
```diff
@@ -18,7 +18,8 @@ Next to `numpy`, `scipy`, and `matplotlib` that are included in the standard
 
-    $ pip install pandas mne
+    $ pip install pandas
+    $ pip install -U https://api.github.com/repos/mne-tools/mne-python/zipball/master
```
maybe we can note the mne-python current version as the minimal in the text above

jasmainak replied:
not sure what you mean?

teonbrooks replied:
I guess we don't specify anywhere minimal version number. like, we should have a section that says mne >= 0.17.X

jasmainak replied:
That's probably because we don't know them. I know only for `MNE` we need >=0.17.x. For the rest I would say the "latest" :-)

teonbrooks commented on examples/convert_eeg_to_bids.py:
```diff
@@ -23,10 +23,11 @@
+import mne
 from mne.datasets import eegbci
-from mne.io import read_raw_edf
+from mne.io.edf.edf import read_annotations_edf
```
this is an mne-python-related question.

jasmainak replied:
yeah, this is related to an issue I raised on `mne-python`: https://github.com/mne-tools/mne-python/issues/5670

teonbrooks commented on examples/convert_eeg_to_bids.py:
```diff
@@ -23,10 +23,11 @@
+from mne.io.edf.edf import read_annotations_edf
 
-from mne_bids import raw_to_bids
+from mne_bids import write_raw_bids, make_bids_basename
```
I am not convinced that we need to expose the `make_bids_basename` to users. I think having users specify subject, run, task, and optionally session in the `write_raw_bids` is probably a better approach

teonbrooks commented on mne_bids/commands/mne_bids_raw_to_bids.py:
```diff
@@ -35,6 +35,9 @@ def run():
+    parser.add_option('--acq', dest='acq',
+                      help='The acquisition parameter.',
+                      metavar='acq')
     parser.add_option('--kind', dest='kind',
```
could you update https://github.com/mne-tools/mne-bids/pull/106/files#diff-0b05a5040f501fe3b6bc1f14bb15c31bL20 `parser.set_defaults(kind='meg')` --> `parser.set_defaults(kind='auto')`

teonbrooks commented on mne_bids/io.py:
```diff
@@ -44,38 +50,18 @@ def _read_raw(raw_fname, electrode=None, hsp=None, hpi=None, config=None,
-    # Neuroscan .cnt format
-    elif ext == '.cnt':
-        raw = io.read_raw_cnt(raw_fname, montage=montage)
+    elif ext in ['.edf', '.bdf']:
```
could you add a TODO note above to make this preload=False and integrated above after the annotations and stim channel drop happens in MNE-Python

jasmainak replied:
So, after reading the MNE-Python code, I realized that `read_edf_annotations` actually loads the entire file into memory before parsing the annotations. Going by @agramfort 's comment [here](https://github.com/mne-tools/mne-python/issues/5670#issuecomment-434446818), it doesn't look like it's happening any time soon. But I can add a TODO note sure :)

teonbrooks commented on mne_bids/pick.py:
```diff
@@ -27,8 +27,6 @@ def get_coil_types():
                                FIFF.FIFFV_COIL_VV_PLANAR_T2,
                                FIFF.FIFFV_COIL_VV_PLANAR_T3),
                 megmag=(FIFF.FIFFV_COIL_POINT_MAGNETOMETER,
-                        FIFF.FIFFV_COIL_POINT_MAGNETOMETER_X,
```
I assume these are used?

jasmainak replied:
yep, this is related to #119

jasmainak commented:
> I don't think that make_bids_basename should be exposed to the end user and that we should incorporate the arguments into write_raw_bids and privatize to _make_bids_basename

So, we have an example for this function [here](https://mne-tools.github.io/mne-bids/auto_examples/create_bids_folder.html#sphx-glr-auto-examples-create-bids-folder-py). Do you envision this example going away if it becomes private? 

I would just let the user create the BIDS filename using string formatting but if you think that's adding friction, we can also have it as function arguments. No strong opinion there. My only concern is that since BIDS is still evolving, it might make sense not to tie API decisions to the file naming structure. There are still key-value pairs being added and dropped ...

teonbrooks commented:
> There are still key-value pairs being added and dropped ...

grr... yeah, does this include all current applicable key-value pairs? if they are exposed within the function as optional, then I would actually be ok with having this as a separate function. 

and yes, i do realize i was the one who suggested the basename function and then critique it afterwards lol ;)

jasmainak commented:
> grr... yeah, does this include all current applicable key-value pairs?

yep exactly. You would need to give people a way to write filenames with _all_ the key-value pairs in BIDS, not just a subset ...

teonbrooks commented on mne_bids/utils.py:
```diff
@@ -62,14 +62,51 @@ def _mkdir_p(path, overwrite=False, verbose=False):
+    return kind
+
+
+def make_bids_basename(subject=None, session=None, task=None,
```
nit: `prefix` arg before `suffix`

teonbrooks commented on mne_bids/mne_bids.py:
```diff
@@ -528,32 +517,39 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
-    raw_file : str | instance of mne.Raw
-        The raw data. If a string, it is assumed to be the path to the raw data
-        file. Otherwise it must be an instance of mne.Raw
+    raw : instance of mne.Raw
```
for raw, you can pass the filename, correct? the docstring should be updated to reflect that

jasmainak replied:
oh no, just a `raw` instance. If you pass filename then you would also have to give all the params for the reader ...

teonbrooks commented on mne_bids/mne_bids.py:
```diff
@@ -528,32 +517,39 @@ def _sidecar_json(raw, task, manufacturer, fname, kind, eeg_ref=None,
+    raw : instance of mne.Raw
+        The raw data. It must be an instance of mne.Raw. The data should not be
+        loaded on disk, i.e., raw.preload must be False.
+    bids_fname : str
```
this should be `bids_basename`

teonbrooks commented on mne_bids/mne_bids.py:
```diff
@@ -562,22 +558,8 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
         The event id dict used to create a 'trial_type' column in events.tsv
     hpi : None | str
```
are we keeping `hpi` in this function call? I think that should be handled in `make_bids_basename`

jasmainak replied:
Good catch. I think we can now make use of `raw._init_kwargs` ;-) It's very nifty ...

agramfort commented:
in make_bids_folders you use root for what you call output_path elsewhere no?

jasmainak commented:
Good catch. Now I made the variable `output_path` consistently everywhere

teonbrooks commented:
everything looks good here now and all tests passing. merging.
thanks heaps @jasmainak!
